File: bench/stress/s8_canonical.go
// S8 — canonical k-mer spectrum, Go (std only). go run s8_canonical.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

const K = 10

func revcomp(code uint64, k int) uint64 {
	var rc uint64
	for i := 0; i < k; i++ {
		rc = (rc << 2) | (3 - (code & 3))
		code >>= 2
	}
	return rc
}

func main() {
	f, _ := os.Open("data/genome.fa")
	defer f.Close()
	seq := make([]byte, 0, 1<<20)
	sc := bufio.NewScanner(f)
	sc.Buffer(make([]byte, 1<<20), 1<<20)
	for sc.Scan() {
		line := sc.Bytes()
		if len(line) > 0 && line[0] == '>' {
			continue
		}
		seq = append(seq, line...)
	}

	var mask uint64 = (1 << (2 * K)) - 1
	var code uint64
	counts := make(map[uint64]uint32)
	for i, b := range seq {
		var two uint64
		switch b {
		case 'A':
			two = 0
		case 'C':
			two = 1
		case 'G':
			two = 2
		case 'T':
			two = 3
		default:
			continue
		}
		code = ((code << 2) | two) & mask
		if i+1 >= K {
			rc := revcomp(code, K)
			canon := code
			if rc < canon {
				canon = rc
			}
			counts[canon]++
		}
	}

	var total uint64
	var max uint32
	for _, c := range counts {
		total += uint64(c)
		if c > max {
			max = c
		}
	}
	fmt.Printf("distinct=%d\n", len(counts))
	fmt.Printf("total=%d\n", total)
	fmt.Printf("max=%d\n", max)
}

File: bench/stress/s8_canonical.rs
// S8 — canonical k-mer spectrum, Rust (std only). 2-bit rolling code; the reverse
// complement is computed on the packed code (complement = base XOR 3, reversed),
// and canonical = min(code, revcomp_code) — which equals the lexicographically
// smaller string because the code is big-endian (first base = high bits).
// rustc -O s8_canonical.rs -o s8_canonical
use std::collections::HashMap;
use std::fs;

const K: usize = 10;

fn revcomp(mut code: u64, k: usize) -> u64 {
    let mut rc: u64 = 0;
    for _ in 0..k {
        rc = (rc << 2) | (3 - (code & 3)); // complement of a 2-bit base = 3 - base
        code >>= 2;
    }
    rc
}

fn main() {
    let raw = fs::read_to_string("data/genome.fa").unwrap();
    let mut seq: Vec<u8> = Vec::with_capacity(raw.len());
    for line in raw.lines() {
        if !line.starts_with('>') {
            seq.extend(line.bytes());
        }
    }

    let mask: u64 = (1u64 << (2 * K)) - 1;
    let mut code: u64 = 0;
    let mut counts: HashMap<u64, u32> = HashMap::new();
    for (i, &b) in seq.iter().enumerate() {
        let two = match b {
            b'A' => 0u64,
            b'C' => 1,
            b'G' => 2,
            b'T' => 3,
            _ => continue,
        };
        code = ((code << 2) | two) & mask;
        if i + 1 >= K {
            let rc = revcomp(code, K);
            let canon = code.min(rc);
            *counts.entry(canon).or_insert(0) += 1;
        }
    }

    let total: u64 = counts.values().map(|&c| c as u64).sum();
    let max = counts.values().copied().max().unwrap_or(0);
    println!("distinct={}", counts.len());
    println!("total={}", total);
    println!("max={}", max);
}
